pkg/sanity/controller.go

```diff
@@ -1580,18 +1516,135 @@ var _ = DescribeSanity("ExpandVolume [Controller Server]", func(sc *TestContext)
 		}
-		rsp, err := r.ControllerExpandVolume(context.Background(), expReq)
+		rsp, err := r.ControllerModifyVolume(context.Background(), modifyReq)
+		ExpectErrorCode(rsp, err, codes.InvalidArgument)
+	})
+
+	// TODO Q: Should we enforce Empty MutableParameters list as InvalidArgument Err?
+	// TODO Q Cont: KEP says it is OPTIONAL, k/k: storage/types says required, but no mention in CSI Spec
+	It("should fail if no mutable parameters specified", func() {
+
+		By("creating a new volume")
+
+		volReq := MakeCreateVolumeReq(sc, UniqueString("sanity-modify-volume"))
+		vol := r.MustCreateVolume(context.Background(), volReq)
+
+		By("failing to modify the volume")
+
+		modifyReq := &csi.ControllerModifyVolumeRequest{
+			VolumeId: vol.GetVolume().GetVolumeId(),
+			Secrets:  sc.Secrets.ControllerModifyVolumeSecret,
+		}
+		rsp, err := r.ControllerModifyVolume(context.Background(), modifyReq)
+		ExpectErrorCode(rsp, err, codes.InvalidArgument)
+	})
+
+	It("should modify a volume created without a volume attribute class", func() {
+
+		By("creating a new volume")
+
+		volReq := MakeCreateVolumeReq(sc, UniqueString("sanity-modify-volume"))
+		vol := r.MustCreateVolume(context.Background(), volReq)
+
+		By("modifying the volume")
+
+		modifyReq := MakeModifyVolumeReq(sc, vol.GetVolume().GetVolumeId())
+		rsp, err := r.ControllerModifyVolume(context.Background(), modifyReq)
+
 		Expect(err).NotTo(HaveOccurred())
 		Expect(rsp).NotTo(BeNil())
-		Expect(rsp.GetCapacityBytes()).To(Equal(TestVolumeExpandSize(sc)))
+	})
+
+	It("should modify a volume created with a volume attribute class", func() {
+
+		By("creating a new volume with volume attribute class")
+
+		volReq := MakeCreateVolumeReq(sc, UniqueString("sanity-modify-volume"))
+		volReq.MutableParameters = sc.Config.TestVolumeMutableParameters
+		vol := r.MustCreateVolume(context.Background(), volReq)
+
+		By("modifying the volume")
+
+		modifyReq := MakeModifyVolumeReq(sc, vol.GetVolume().GetVolumeId())
+		rsp, err := r.ControllerModifyVolume(context.Background(), modifyReq)
+
+		Expect(err).NotTo(HaveOccurred())
+		Expect(rsp).NotTo(BeNil())
+	})
+
+	It("should fail to modify a volume created with a volume attribute class if new mutable parameters are not supported by volume", func() {
+
+		By("creating a new volume with volume attribute class")
+
+		volReq := MakeCreateVolumeReq(sc, UniqueString("sanity-modify-volume"))
+		volReq.MutableParameters = sc.Config.TestVolumeMutableParameters
+		vol := r.MustCreateVolume(context.Background(), volReq)
+
+		By("failing to modify the volume")
+
+		modifyReq := &csi.ControllerModifyVolumeRequest{
+			VolumeId: vol.GetVolume().GetVolumeId(),
+			MutableParameters: map[string]string{
+				"XXX_FakeKey": "XXX_FakeValue",
+			},
+			Secrets: sc.Secrets.ControllerModifyVolumeSecret,
+		}
+		rsp, err := r.ControllerModifyVolume(context.Background(), modifyReq)
+		ExpectErrorCode(rsp, err, codes.InvalidArgument)
 	})
```

Comment:
In the CSI spec, it says 'Values specified in mutable_parameters MUST take precedence over the values from parameters.`.  Can you add a test case for that?

Reply:
An issue I had implementing this test case pre-pr submission is that `parameters` & `mutable_parameters` are opaque, and we have no visibility into created volume respecting the precedence.

One way we could potentially test this is by finding a shared parameter + mutable_parameter (looping through the two lists), and forcing the mutable_parameter to be invalid. If the driver would return `INVALID_ARGUMENTS` error code, it suggests that precedence is respected, and the test would pass. 

However, I was afraid that this would be too 'hacky' of a test to add, because we don't know if there are shared parameters, or what an invalid parameter might be. 

---

Hemant suggested that the CSI Driver writer can pass in an additional flag where a shared parameter name can be passed, and requiring that this flag be passed if driver implements the ModifyVolume RPC. With this + the invalid_arguments precedence test above we could test this CSI Spec requirement. 

Pros:
- This forces CSI Driver writer to know about this precedence requirement. 

Cons:
- Another test flag/file. Not the most intuitive to add because it likely needs to mention what would make the parameter invalid. 
- We need some way of skipping this test if the tested CSI Driver does not share any parameters between `parameters` and `mutable_parameters`. 

---

I'll go add a commit with this test and additional flag, and then we can debate its inclusion once we see the code.

Reply:
Resolving for now in place of ^^^ issue